### PR TITLE
links-fe: infinite scroll for link displays

### PR DESCRIPTION
**Pull request by liam-fitzgerald**
by request of @vvisigoth 
(nb: this doesn't lazy load the graph, just displays it lazily) 
cc: @tylershuster pls review

### CODE REVIEW

**Comment by vvisigoth** [comment]
big
—
~poldec-tonteg
http://urbit.org

File: pkg/interface/src/views/apps/links/LinkWindow.tsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  api: GlobalApi;
+}
+export function LinkWindow(props: LinkWindowProps) {
+  const { graph, api, association } = props;
+  const loadedNewest = useRef(true);
+  const loadedOldest = useRef(false);
+  const virtualList = useRef<VirtualScroller>();
+  const fetchLinks = useCallback(
+    async (newer: boolean) => {
+      /* stubbed, should we generalize the display of graphs in virtualscroller?
```
Comment (tylershuster) [inline pkg/interface/src/views/apps/links/LinkWindow.tsx:36]: Yes, but maybe not today

File: pkg/interface/src/views/apps/links/LinkResource.tsx [hunk 2]
```diff
@@ -73,27 +74,23 @@ export function LinkResource(props: LinkResourceProps) {
...
-                      baseUrl={resourceUrl}
-                      group={group}
-                      path={resource["group-path"]}
-                      api={api}
-                      mb={3}
-                    />
+                <LinkWindow
+                  association={resource}
+                  contacts={contacts}
+                  key={date.toString()}
```
Comment (tylershuster) [inline pkg/interface/src/views/apps/links/LinkResource.tsx:80]: This throws an error as date is not defined

File: pkg/interface/src/views/apps/links/LinkResource.tsx [hunk 3]
```diff
@@ -73,27 +74,23 @@ export function LinkResource(props: LinkResourceProps) {
...
+                  resource={resourcePath}
+                  graph={graph}
+                  unreads={unreads}
+                  nickname={contact?.nickname}
```
Comment (tylershuster) [inline pkg/interface/src/views/apps/links/LinkResource.tsx:84]: contact also throws an error

File: pkg/interface/src/views/apps/links/LinkResource.tsx [hunk 4]
```diff
@@ -73,27 +74,23 @@ export function LinkResource(props: LinkResourceProps) {
...
+                  hideNicknames={hideNicknames}
+                  remoteContentPolicy={remoteContentPolicy}
+                  baseUrl={resourceUrl}
+                  group={group}
+                  path={resource["group-path"]}
+                  api={api}
+                  mb={3}
+                />
                   );
                 })}
```
Comment (tylershuster) [inline pkg/interface/src/views/apps/links/LinkResource.tsx:95]: some leftover crud: `); })}`

**Comment by tylershuster** [comment]
Not as wide as it should be:

image

**Comment by tylershuster** [comment]
In general this is good, other than the few comments I made. Seems like an appropriate use of the component. I'm fine shipping this now and reworking `VirtualScroller` to function more generically with a graph later.

**Review by tacryt-socryp — APPROVED** [review]

**Comment by matildepark** [comment]
It isn't loading and unloading embedded content.

image

**Comment by matildepark** [comment]
The rest of the page doesn't load in either. 
image

**Comment by liam-fitzgerald** [comment]
ping @matildepark  

also includes some nice perf improvements for virtualscroller now

**Comment by matildepark** [comment]
Weirdly the first page of embeds won't load in but the following ones will? Try this out in Tlon music.

**Comment by matildepark** [comment]
Also getting `“measure is not a function”` when I browse to a LinkItem's page of comments

**Comment by matildepark** [comment]
You may have to banish the ghost in our Portal'd embed logic that unloads all these...

**Comment by liam-fitzgerald** [comment]
uh so the first page not loading in is actually a 403 from youtube, refusing to embed the video. Same issue exists in current landscape, so going to treat that as out of scope for this PR. 

> You may have to banish the ghost in our Portal'd embed logic that unloads all these...

wat mean?

`measure` thing now fixed

**Comment by matildepark** [comment]
Weird. I just assumed we were unloading it permanently.

File: pkg/interface/src/views/components/VirtualScroller.tsx [hunk 5]
```diff
@@ -309,15 +302,15 @@ export default class VirtualScroller extends Component<VirtualScrollerProps, Vir
             height: element.offsetHeight,
             element
           });
-          _.debounce(this.recalculateTotalHeight, 500)();
+          this.recalculateTotalHeight();
```
Comment (tylershuster) [inline pkg/interface/src/views/components/VirtualScroller.tsx:305]: Oh shoot, this is where the debouncer was getting duplicated, huh?

Comment (liam-fitzgerald) [reply pkg/interface/src/views/components/VirtualScroller.tsx:305]: Yeah, perf gains are actually pretty nice

**Comment by matildepark** [comment]
I honestly am having a hard time judging the performance improvements here. It seems to be the same speed and firing roughly as often when I compile in `production`?

**Comment by matildepark** [comment]
You can definitely scroll further up without things getting clunky. And while we are getting spotty embeds of YouTube in Links displays, it appears to precede this PR as an issue, and it's such a strict improvement over the current "load all 500 YouTube vids" that I'm fine with this PR as is. Will merge after CI.

**Review by matildepark — APPROVED** [review]

**Pull request merged**